cluster-autoscaler/core/scale_down.go

```diff
@@ -403,14 +403,27 @@ func (sd *ScaleDown) UpdateUnneededNodes(
 
 		if err != nil {
 			klog.Warningf("Failed to calculate utilization for %s: %v", node.Name, err)
+
+			// Skip node that GPU resource is not ready to avoid scaling down new created node.
+			if gpu.NodeHasGpu(node) {
```

Comment:
I understand what problem you're solving with this, but this feels like a very hacky way to do this. 
We consider node with GPU not ready as NotStarted in clusterstate. The correct thing to do would be to have proper handling for not started nodes (don't scale them down, delete them after timeout). 
Slightly less correct thing would be an explicit if checking for node with unready GPU and skipping it in there.

Reply:
Oh. I should leave a comment here. Originally, I think it doesn't need extra handling here. For node with GPU not ready case, utilization will be 0 and it will be added to candidate collection. Ideally, GPU is ready before `ScaleDownUnneededTime`. Next, CA will reconsider this node based on GPU usage. Do you think that's reasonable?  Delete line 408-410

Reply:
But this is still an explicit way..

Reply:
@MaciekPytel. I am trying to add logic to explicitly handle unready GPU Nodes. A new collection is needed to track those nodes. 
1. The `delete` you mean is delete kubernetes node and delete them in cloudprovider? Like `removeOldUnregisteredNodes` . The difference here is unready GPU nodes may have workloads running after timeout period unless user taints and tolerance or node affinity. If there's workload, better to scale them down.
2.  Does CA already have a timeout to use here? I searched configs and only find `unremovableNodeRecheckTimeout`. It probably need another option, like 5 or 10?

Reply:
The current implementation will never remove a node with unready GPU. There should be some timeout for device plugin after which the node will be deleted (to avoid keeping defunct costly nodes around). Have you actually tested what will happen with those nodes? I suspect they will be deleted after `--scale-down-unready-time` (which is not perfect, but at least possible to change with a flag)?

Reply:
@MaciekPytel No, the case device plugin failed to report has not been covered in my test. This is a good edge case. In current implementation, if we just consider gpu utilization, calculator will return error here and node will be added into unneededList in scale down phase1. 

Since node status is still `ready`, it will be treated as ready node in phase 2. 

I think another option is
1. Add `FilterOutNodesWithUnreadyGpus` at the beginning of phase1, this will make sure CA won't consider unReady GPU nodes, this won't cover the case we want to delete nodes after 

2. Add hack logic here
https://github.com/kubernetes/autoscaler/blob/bd84757b7eedee6e122194ef7a1f41af27e45783/cluster-autoscaler/core/scale_down.go#L656

mark unready gpu node `unready` in phase 2, force the case go through 
https://github.com/kubernetes/autoscaler/blob/bd84757b7eedee6e122194ef7a1f41af27e45783/cluster-autoscaler/core/scale_down.go#L664-L666

Reply:
Hi @MaciekPytel  I double verified behaviors in tests. The problem I think is no matter node is ready or unready, CA calculates utilization for the node and add it into `unneededNodeList`.  

In this case, we assume node is unready, 
1. If device plugin is working fine, after a while (normally before ScaleDownUnneededTime), the node becomes ready. If utilization rate is low, it will be added into `unneededNodeList` 

2. If device plugin isn't working well, node will be unready and it's added into `unneededNodeList`.  CA will wait ScaleDownUnreadyTime to delete node. 


I think if we want to explicitly manage unready nodes, a reasonable solution is to skip unready node and only add those nodes to unneededNostList after `--scale-down-unready-time` in Phase I. Doesn't need to calculate utilization for them.  Need a map similar like `unneededNodes` to track those nodes.

Reply:
If it's working well, why do we need to introduce the new map? Seems like all it's doing is save us the utilization check, which is basically free. The actual behavior would stay the same with your change, we've just added extra map and more code to handle it.

Reply:
@MaciekPytel  I think last time, you mentioned it's better to `explicitly handle unready GPU Nodes`. Maybe I am doing the wrong way? 

The original way was a little bit tricky. Every nodes get into resource calculation but new coming GPU node will fail calculation and return error. Then we check if it's gpu case, we ignore it until node get allocatable gpu.  Do you think this is fine? I can change back to this way. It's working well but not that elegant.

The latest change brings new map and behavior won't change, just explicitly filter unready nodes before going to utilization calculation phase. But it has to control time because it could never come back and then we want to remove it.

Reply:
I think I misunderstood what you were trying to do with that error handling - I thought you wanted to change what happens if driver installation falls and now I think it's just about handling missing Allocatable in calculateUtilizationOfResource. Is that correct?

Reply:
Yes. Just as you said. I have not seen any situation driver installation falls yet so I didn't consider this case at that time.

Reply:
Right, so I indeed misunderstood. I'm thinking maybe the best way to handle this would be to just handle it in calculateUtilizationOfResource (make it return 0 if GPU utilization is missing)? The current solution feels overly complicated and I don't like the fact that previous one would ignore any error, not just this specific one.

Reply:
I see. Let me have a refactor and try to simply this change within small scope. Once I am done, I will come to you for another review.